### ISSUE TITLE
Overlay .devcontiner

### ISSUE THREAD

**Issue by bhack** [microsoft/vscode-remote-release#3279]
We need to find a way to overlay `.devcontainer`.
Suppose that user are switch commented lines on default setting to enable GPU as in https://github.com/microsoft/vscode-dev-containers/pull/381. 
I still want to distribute updates but if users are editing the same file that I am distributing  it will be hard to get updates. So I suppose that user override need to be placed in a separate file. 
Just to explore some solutions we could get some ideas from https://github.com/kubernetes-sigs/kustomize#2-create-variants-using-overlays.

**Comment by Chuxel** [comment]
While not directly what you are mentioning, this may be of interest: https://github.com/microsoft/vscode-remote-release/issues/3218. You'll soon be able to add your own definitions to for repos locally.

//cc: @chrmarti

**Comment by bhack** [comment]
Are you talking about global override or overlay (partial ovveride)?

**Comment by Chuxel** [comment]
@bhack It is a global override not an overlay.

**Comment by bhack** [comment]
Yes here It more like I want to be update with the distributed configurations but I work with some small local changes (e.g. an extra mount, an extra env var etc..) or with some lines that was distributed commented (i.e. default CPU but I an on a GPU machine).

**Comment by WasabiFan** [comment]
I'd just like to bump this thread and share my use-case for something along these lines.

In my case, I would like to add an extra mount in the dev container for my personal use which doesn't affect the repository copy. I don't want to .gitignore the `.devcontainer.json` since that is a shared file which everyone edits. As far as I can tell, if you have a `.devcontainer.json` checked into your repo, there isn't a supported way to have user-specific configuration on top of that base container definition. It would make sense to me if you could add those in your user configuration, or add a `.devcontainer-user.json` which _was_ expected to be gitignored.

**Comment by bhack** [comment]
Any progress on this?

**Comment by chrmarti** [comment]
If we want to overlay 2 JSON files, we need to consider the options for merging objects and arrays. Overwriting top-level properties might not always be the desired behavior, while merging properties (or adding elements) recursively excludes the option for removing properties (or elements).

**Comment by bhack** [comment]
We need something like https://kubernetes.io/docs/tasks/manage-kubernetes-objects/kustomization/#bases-and-overlays

**Comment by Chuxel** [comment]
@chrmarti Yeah this in my view is highly related to https://github.com/microsoft/vscode-remote-release/issues/3568. We ultimately need an inheritence model, ideally one that can use some sort of central registry as well.  With what @bhack has pointed out, and that we are seeing people placing dev container config in sub-folders, you can imagine a scenario where there is a GPU and CPU flavor of the config... if not other more dramatic differences.  

If `.devcontainer.json` could be added to `.gitignore`, you could store a full `.devcontainer` folder somewhere else - say `${workspaceFolder}/dev/gpu`.  This `.devcontainer.json` file could then just include something like: 

```
{
    "parent": "${workspaceFolder}/dev/gpu"
}
```

I've also heard about scenario driven dev container definitions that this would also work for (e.g. you're in front-end dev mode verses back end dev mode).

Allowing devcontainer.json properties to be overidden or merged, then becomes a natural extension.  We'd need to determine property behaviors, but to some extent its a bit like having `docker-compose.yml` and then `docker-compose.override.yml` with tweaks.  Certainly `build.args` is a good example of one I'd exepct to be of interest particularly if the dev container parent comes from somewhere else.

**Comment by chrmarti** [comment]
`docker-compose.yml` doesn't allow for removing array elements / replacing arrays or deleting object properties: https://docs.docker.com/compose/extends/#adding-and-overriding-configuration

I'd be wary of the complexity we introduce if it doesn't solve more than the simple cases. E.g., Webpack and ESBuild use JavaScript to cover all cases. Could we do the same and have a `devcontainer.js`? (This would assume that we will have a JavaScript runtime available whenever we need to evaluate the configuration, but maybe that's a reasonable assumption?)

**Comment by bhack** [comment]
It hard to develop vs distribute `.devcontainer` in a github repo as you really don't know if you want to put the `.devcontainer` folder under `.gitignore` when you need to customize it.

**Comment by guye1296** [comment]
Any updates on this?

**Comment by ian-h-chamberlain** [comment]
In lieu of this feature, I had been using the precedence mentioned in https://github.com/microsoft/vscode-remote-release/issues/3218#issuecomment-646636472 to do something like this workflow:

* Create a directory like `~/.config/devcontainers`
* Configure the new directory in `dev.containers.repositoryConfigurationPaths`
* Copy a checked-out repo's `.devcontainer/devcontainer.json` to `~/.config/devcontainers/github.com/$ORG/$REPO/.devcontainer/`
* Customize the version in `~/.config/devcontainers` as needed
* `Rebuild and Reopen in Container`. The precedence mentioned above ensured my customized version gets used

However, this workflow seems to have stopped working — I'm not sure when, unfortunately I don't recreate my containers very often. Now, the checked-in `devcontainer.json` seems to take precedence over my local one (as initially proposed in https://github.com/microsoft/vscode-remote-release/issues/3218#issuecomment-646608062). I can see from the logs that the `label=devcontainer.config_file` refers to the repository version, rather than my local `~/.config` file.

Has anyone else noticed this change / should I file a new issue for it? Without this workflow I can't seem to come up with any way at all to customize my local container, other than deleting the repository's `.devcontainer` folder every time I want to use it (obviously a huge pain!).

**Comment by 99linesofcode** [comment]
I specifically like the suggestion as proposed in #22 and look very much forward to this becoming part of the official spec.

**Comment by gamesover** [comment]
Any updates? Ideally, we should have a mechanism similar to `.env` and `.env.local`, where the local file can override the default configuration.

**Comment by nandyalu** [comment]
Any updates on this? I would really like to have something like `devcontainer.json` with basic configuration for the repo devcontainer, and then let devs have their own `devcontainer.local.json` with their specific configuration as proposed https://github.com/microsoft/vscode-remote-release/issues/7650#issuecomment-1380146457 and add it to `gitignore` so that it remains local and individually customizable.

**Comment by chrmarti** [comment]
Not exactly what you are asking for, but in some cases this might work: You can use the Dev Containers CLI to build an image from a first devcontainers.json (there is also a GitHub Action for this) and then use that image in a second devcontainers.json to further customize it. The first image includes the configuration from the first devcontainer.json that will be merged with what is in the second devcontainer.json.

**Comment by nandyalu** [comment]
So, prebuild a devcontainer image using an action like [this](https://github.com/devcontainers/ci/blob/main/docs/github-action.md#other-examples) and then create local devcontainer.json with user settings and use the base image as the prebuilt image?

Do you happen to have any examples I can look at?

Edit: I was looking at Github Action in the link and that doesn't seem like the action we need. Ignore the link, any other details on how I can test/implement this would be appreciated!

**Comment by chrmarti** [comment]
https://github.com/devcontainers/ci can be used to build an image in CI. You can do the same locally with the Dev Containers CLI: https://github.com/devcontainers/cli.

E.g.: Have a `.devcontainer/base/devcontainer.json` for the base image and a `.devcontainer/devcontainer.json` for development. You can then build the base image with:
```sh
devcontainer build --workspace-folder . --config .devcontainer/base/devcontainer.json --image-name project-base-image:latest
```

The `.devcontainer/devcontainer.json` can then refer to the base image by the name provided with `--image-name`.

The GitHub action then allows you to automate this step and push it to an image registry to make it available to others.

**Comment by nandyalu** [comment]
https://github.com/devcontainers/cli This is a helpful link... I will give this a try! Thank you
